phpBB/phpbb/filesystem/filesystem.php

```diff
@@ -0,0 +1,879 @@
+ */
+class filesystem implements filesystem_interface
+{
+	/**
+	 * Store some information about file ownership for phpBB's chmod function
+	 *
+	 * @var array
+	 */
+	protected $chmod_info;
+
+	/**
+	 * Stores current working directory
+	 *
+	 * @var string|bool		current working directory or false if it cannot be recovered
+	 */
+	protected $working_directory;
+
+	/**
+	 * Symfony's Filesystem component
+	 *
+	 * @var \Symfony\Component\Filesystem\Filesystem
+	 */
+	protected $symfony_filesystem;
+
+	/**
+	 * Constructor
+	 */
+	public function __construct()
+	{
+		$this->chmod_info			= array();
+		$this->symfony_filesystem	= new \Symfony\Component\Filesystem\Filesystem();
+		$this->working_directory	= null;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function chgrp($files, $group, $recursive = false)
+	{
+		try
+		{
+			$this->symfony_filesystem->chgrp($files, $group, $recursive);
+		}
+		catch (\Symfony\Component\Filesystem\Exception\IOException $e)
+		{
+			// Try to recover filename
+			// By the time this is written that is at the end of the message
+			$error = trim($e->getMessage());
+			$file = substr($error, strrpos($error, ' '));
+
+			throw new filesystem_exception('CANNOT_CHANGE_FILE_GROUP', $file, array(), $e);
+		}
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function chmod($files, $perms = null, $recursive = false)
+	{
+		if (is_null($perms))
+		{
+			// Default to read permission for compatibility reasons
+			$perms = self::CHMOD_READ;
+		}
+
+		// Check if we got a permission flag
+		if ($perms > self::CHMOD_ALL)
+		{
+			$file_perm = $perms;
+
+			// Extract permissions
+			//$owner = ($file_perm >> 6) & 7; // This will be ignored
+			$group = ($file_perm >> 3) & 7;
+			$other = ($file_perm >> 0) & 7;
+
+			// Does any permissions provided? if so we add execute bit for directories
+			$group = ($group !== 0) ? ($group | self::CHMOD_EXECUTE) : $group;
+			$other = ($other !== 0) ? ($other | self::CHMOD_EXECUTE) : $other;
+
+			// Compute directory permissions
+			$dir_perm = (self::CHMOD_ALL << 6) + ($group << 3) + ($other << 3);
+		}
+		else
+		{
+			// Add execute bit to owner if execute bit is among perms
+			$owner_perm	= (self::CHMOD_READ | self::CHMOD_WRITE) | ($perms & self::CHMOD_EXECUTE);
+			$file_perm	= ($owner_perm << 6) + ($perms << 3) + ($perms << 0);
+
+			// Compute directory permissions
+			$perm = ($perms !== 0) ? ($perms | self::CHMOD_EXECUTE) : $perms;
+			$dir_perm = (($owner_perm | self::CHMOD_EXECUTE) << 6) + ($perm << 3) + ($perm << 0);
+		}
+
+		// Symfony's filesystem component does not support extra execution flags on directories
+		// so we need to implement it again
+		foreach ($this->to_iterator($files) as $file)
+		{
+			if ($recursive && is_dir($file) && !is_link($file))
+			{
+				$this->chmod(new \FilesystemIterator($file), $perms, true);
+			}
+
+			// Use directory permissions on symlinks as well
+			if (is_dir($file) || is_link($file))
```

Comment:
Not sure if chmoding is a good idea, it works on macs for sure, but most linux systems will result in error as links only allowed with 0777.